Remotion/Data/DomainObjects/Persistence/ReadOnlyStorageProviderDecorator.cs

```diff
@@ -0,0 +1,125 @@
+  /// Decorator for <see cref="IReadOnlyStorageProvider"/> to ensure that <see cref="IQuery"/>s have the appropriate <see cref="QueryType"/> when
+  /// being executed in the ReadOnly context.
+  /// </summary>
+  public class ReadOnlyStorageProviderDecorator : IReadOnlyStorageProvider
+  {
+    public ReadOnlyStorageProviderDecorator (IReadOnlyStorageProvider innerStorageProvider)
+    {
+      ArgumentUtility.CheckNotNull("innerStorageProvider", innerStorageProvider);
+
+      InnerStorageProvider = innerStorageProvider;
+    }
+
+    public IReadOnlyStorageProvider InnerStorageProvider { get; }
+
+    public void Dispose ()
+    {
+      InnerStorageProvider.Dispose();
+    }
+
+    public ObjectLookupResult<DataContainer> LoadDataContainer (ObjectID id)
+    {
+      ArgumentUtility.CheckNotNull("id", id);
+
+      return InnerStorageProvider.LoadDataContainer(id);
+    }
+
+    public IEnumerable<ObjectLookupResult<DataContainer>> LoadDataContainers (IReadOnlyCollection<ObjectID> ids)
+    {
+      ArgumentUtility.CheckNotNull("ids", ids);
+
+      return InnerStorageProvider.LoadDataContainers(ids);
+    }
+
+    public IEnumerable<DataContainer> LoadDataContainersByRelatedID (
+        RelationEndPointDefinition relationEndPointDefinition,
+        SortExpressionDefinition? sortExpressionDefinition,
+        ObjectID relatedID)
+    {
+      ArgumentUtility.CheckNotNull("relationEndPointDefinition", relationEndPointDefinition);
+      ArgumentUtility.CheckNotNull("relatedID", relatedID);
+
+      return InnerStorageProvider.LoadDataContainersByRelatedID(relationEndPointDefinition, sortExpressionDefinition, relatedID);
+    }
+
+    public IEnumerable<DataContainer?> ExecuteCollectionQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteCollectionQuery(query);
+    }
+
+    public IEnumerable<IQueryResultRow> ExecuteCustomQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteCustomQuery(query);
+    }
+
+    public object? ExecuteScalarQuery (IQuery query)
+    {
+      ArgumentUtility.CheckNotNull("query", query);
+
+      CheckQueryType(query);
+
+      return InnerStorageProvider.ExecuteScalarQuery(query);
+    }
+
+    public void BeginTransaction ()
+    {
+      InnerStorageProvider.BeginTransaction();
+    }
+
+    public void Commit ()
+    {
+      InnerStorageProvider.Commit();
+    }
+
+    public void Rollback ()
+    {
+      InnerStorageProvider.Rollback();
+    }
+
+    private void CheckQueryType (IQuery query)
+    {
+      _ = query.QueryType switch
+      {
+        QueryType.CollectionReadWrite or QueryType.ScalarReadWrite or QueryType.CustomReadWrite => throw new ArgumentException(
+            $"Read-write query '{query}' cannot be used with IReadOnlyStorageProvider.", nameof(query)),
+        _ => true
```

Comment:
I think this would read better as an IF statement. The switch-syntax without actually using the result values makes it very difficult to understand what the actual intent is without already knowing it.

We should consider giving more context with the ArgumentException:
"Query '{query.ID}' has query type '{query.Type}' which cannot be used in a read-only query scenario."